File: iOS App/MasterKey/MasterKey/Models/sha256.swift
//
//  sha256.swift
//  MasterKey
//
//  Created by Wembley on 27/4/22.
//

import Foundation
import CryptoKit

//Clase dedicada a retornar un hash resultante en SHA256
class sha256 {
    
    //Metodo estatico que recibe una cadena y devuelve su hash a traves del algoritmo SHA256
    static func encrypt(message: String) -> String {
        //Creamos un objeto data a partir del mensaje
        let inputData = Data(message.utf8)
        //Obtenemos el hash sha256 de la cadena
        let hashed = SHA256.hash(data: inputData)
        //Obtenemos el formato de cadena hexadecimal del hash
        let hashedString = hashed.compactMap { String(format: "%02x", $0) }.joined()

        return hashedString //retorna el hash SHA256 resultante de la cadena recibida por parametro
    }
}

File: Desktop App/MasterKey/src/EncryptionModels/SHA256.java
package EncryptionModels;

import java.security.MessageDigest;

/**
 * @autor: Alvaro
 */
public class SHA256 {

    /*Clase dedicada a retornar hashes a partir de Strings.
    Usa el algotitmo de cifrado hash SHA-256, que pertenece al conjunto de SHA-2 el cual fue diseñado por la NSA.
    Su función es crear una cadena resumen (digest) del texto original (En este caso la password).
    Este valor resultante tiene una longitud fija y no se puede obtener la cadena original, ni su longitud a partir de este.
    Esta función hash está pensada para que sea posible una encriptacion muy consistente,
    y una dificil desencriptacion, por lo que para encriptar las passwords de la app en la BBDD es idonea.
    (Sólo es posible desencriptar las cadenas comparando hashes (DIFICIL TAREA)).
    PARA ELLO USO EL PAQUETE: java.security*/
    public SHA256() {
    }

    //MEtodo que recoge una cadena y a traves del algoritmo SHA-256 retorna su hash.
    public static String encrypt(String message) {
        try {
            //Creo una instancia de Message Diggest, clase que me permite recoger 
            //el algoritmo de codificacion y realizar metodos con él.
            MessageDigest md = MessageDigest.getInstance("SHA-256");

            //Recojo el resultado del algoritmo en un array de bytes.
            byte[] hashedBytesArray = md.digest(message.getBytes("UTF-8"));

            //Creo un StringBuffer para recoger la password encriptada usando su funcion append().
            StringBuffer hexString = new StringBuffer();

            //Añado al String cada byte del array de bytes anterior, obteniendo la cadena codificada de nuestra password.
            for (int i = 0; i < hashedBytesArray.length; i++) {
                hexString.append(String.format("%02X", hashedBytesArray[i])); //Añado cada byte con String.format(x) x=hexadecimal
            }

            return hexString.toString(); //Retorna el hash SHA-256 del mensaje.

        } catch (Exception e) {
            return null;
        }
    }
}
